[SPARK-9312][ML] Add RawPrediction, numClasses, and numFeatures for OneVsRestModel

lu-wang-dl opened the pull request:
add RawPrediction as output column 
add numClasses and numFeatures to OneVsRestModel

## What changes were proposed in this pull request?

- Add two val numClasses and numFeatures in OneVsRestModel so that we can inherit from Classifier in the future

- Add rawPrediction output column in transform, the prediction label in calculated by the rawPrediciton like raw2prediction

 

## How was this patch tested?


(Please explain how this patch was tested. E.g. unit tests, integration tests, manual tests)
(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)
Please review http://spark.apache.org/contributing.html before opening a pull request.

lu-wang-dl committed:
add rawPrediction as an output column;
add numCLasses and numFeatures to OneVsRestModel

AmplabJenkins commented:
Can one of the admins verify this patch?

SparkQA commented:
**[Test build #89216 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89216/testReport)** for PR 21044 at commit [`0cfc20a`](https://github.com/apache/spark/commit/0cfc20a3637c06071e6fe48ca5db4834b34c889e).

SparkQA commented:
**[Test build #89216 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89216/testReport)** for PR 21044 at commit [`0cfc20a`](https://github.com/apache/spark/commit/0cfc20a3637c06071e6fe48ca5db4834b34c889e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89216/
Test PASSed.

jkbradley commented:
Thanks for the PR!  Quick high-level comment: We'll need to have rawPredictionCol be optional.  If it's not set or is an empty string, then it should not be added to the output DataFrame.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -195,14 +205,18 @@ final class OneVsRestModel private[ml] (
-      predictions.maxBy(_._2)._1.toDouble
+    // output the RawPrediction as vector
+    val rawPredictionUDF = udf { (predictions: Map[Int, Double]) =>
+      Vectors.sparse(numClasses, predictions.toList )
```
Also, let's output a dense Vector since it will almost surely be dense.

lu-wang-dl committed:
make rawPrediction optionall

SparkQA commented:
**[Test build #89308 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89308/testReport)** for PR 21044 at commit [`2a47e2b`](https://github.com/apache/spark/commit/2a47e2be30d52e3fbea7e1eeeaa5048a6ac97116).

lu-wang-dl committed:
change the tmp array name

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89308/
Test PASSed.

SparkQA commented:
**[Test build #89309 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89309/testReport)** for PR 21044 at commit [`0c32fca`](https://github.com/apache/spark/commit/0c32fcaaf87f1922170e4ce7e60381ccd23ab6e8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89309/
Test PASSed.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -138,6 +138,12 @@ final class OneVsRestModel private[ml] (
     @Since("1.4.0") val models: Array[_ <: ClassificationModel[_, _]])
   extends Model[OneVsRestModel] with OneVsRestParams with MLWritable {
 
```
Let's add a require() statement here which checks that models.nonEmpty is true (to throw an exception upon construction, rather than when numFeatures calls models.head below).  Just to be safe...

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -146,6 +152,10 @@ final class OneVsRestModel private[ml] (
 
+  /** @group setParam */
+  @Since("2.4.0")
+  def setRawPredictionCol(value: String): this.type = set(rawPredictionCol, value)
```
You'll need to add this to the Estimator too.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -195,15 +206,32 @@ final class OneVsRestModel private[ml] (
+    // output the RawPrediction as vector
+    if (getRawPredictionCol != "") {
+      val rawPredictionUDF = udf { (predictions: Map[Int, Double]) =>
+        val predArray = Array.fill[Double](numClasses)(0.0)
```
This causes a subtle ContextCleaner bug: `numClasses` refers to a field of the class OneVsRestModel, so when Spark's closure capture serializes this UDF to send to executors, it will end up sending the entire OneVsRestModel object, rather than just the value for numClasses.  Make a local copy of the value numClasses within the transform() method to avoid this issue.

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -195,15 +206,32 @@ final class OneVsRestModel private[ml] (
+        .withColumn(getPredictionCol, labelUDF(col(getRawPredictionCol)), labelMetadata)
+        .drop(accColName)
+    }
+    else {
```
Scala style: This should go on the previous line: ```} else {```

jkbradley commented on mllib/src/main/scala/org/apache/spark/ml/classification/OneVsRest.scala:
```diff
@@ -195,15 +206,32 @@ final class OneVsRestModel private[ml] (
+      val labelUDF = udf { (predictions: Map[Int, Double]) =>
+        predictions.maxBy(_._2)._1.toDouble
+      }
+      // output confidence as rwa prediction, label and label metadata as prediction
```
This comment seems to be in the wrong part of the code.  Also there's a typo

jkbradley reviewed (commented):
Thanks for the PR!  Done with more in-depth review pass

lu-wang-dl committed:
fix typos and comments

SparkQA commented:
**[Test build #89353 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89353/testReport)** for PR 21044 at commit [`ebf4a6c`](https://github.com/apache/spark/commit/ebf4a6c155be6a13fb41f492eb2777465d163478).

SparkQA commented:
**[Test build #89353 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89353/testReport)** for PR 21044 at commit [`ebf4a6c`](https://github.com/apache/spark/commit/ebf4a6c155be6a13fb41f492eb2777465d163478).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89353/
Test PASSed.

lu-wang-dl committed:
add require in OneVsRestModel
add setRawPredictionCol in OneVsRest
create a local var numClass to resolve the issue

SparkQA commented:
**[Test build #89361 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/89361/testReport)** for PR 21044 at commit [`b3c7fec`](https://github.com/apache/spark/commit/b3c7fec0fda9056b832d1d35e829e9946218e504).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/89361/
Test PASSed.

jkbradley commented:
LGTM
Merging with master
Thanks!!
